### PR TITLE
lru cache implementation writeup completed

### LINKED ISSUE #3309: LRU Cache Implementation in Javascript
Proposed title of the article
LRU Cache Implementation in C#

Introduction:
In-memory caches such as Memcached and Redis are key-value stores between your application and your data storage. Since the data is held in RAM, it is much faster than typical databases where data is stored on a disk. RAM is more limited than disk, so cache invalidation algorithms such as least recently used (LRU) can help invalidate 'cold' entries and keep 'hot'
data in RAM.

Key takeaways:
At the end of this article the learner will be able to:

Understand what LRU cache is
Data structures used to build an efficient LRU cache algorithm
Implement LRU cache using C#

References:
N/A

### PR DESCRIPTION
**Pull request by erixbiz**
#### Formatting and structure checks
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**
- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Is the article you are submitting an in-depth and unique article? Does it go beyond what is in the official docs and what is covered in other blog sites. See these articles as [examples](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md).
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**
- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) - [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker. We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**
- [x] Have you checked your code runs correctly and you've highlighted all necessary dependencies for installation?
- [x] Are the software programs and packages you're highlighting in your article up to date, using current versions and not deprecated?

#### Topic suggestion this closes

This closes #3309.

### CODE REVIEW

**Comment by erixbiz** [comment]
@geoffrey45 I have managed to create a new PR.

**Comment by WanjaMIKE** [comment]
@erixbiz what issue does this PR fix?

**Comment by geoffrey45** [comment]
Hello @erixbiz Please read and check the pull header checklist accordingly.

**Comment by erixbiz** [comment]
> @erixbiz what issue does this PR fix?

@WanjaMIKE #3309

**Comment by erixbiz** [comment]
@geoffrey45  done so

File: content/articles/lru-cache-implementation-in-javascript/index.md [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+            this.map = new Map(); //this stores the entire array
+
+            //this is boundaries for double Linked List
+            this.head = {};
+            this.tail = {};
+            this.head.next = this.tail; //initialize your double Linked List
+            this.tail.prev = this.head;
+            };
+
+            ```
```
Comment (geoffrey45) [inline content/articles/lru-cache-implementation-in-javascript/index.md:65]: Please use the correct javascript code indentation.

Also, **all** the code blocks should start at the beginning of the line.

File: content/articles/lru-cache-implementation-in-javascript/index.md [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+            this.tail.prev = c; //update last element as tail
+            return c.value;
+            } else {
+            return -1; //element does not exist
+            }
+
+            };
+
+    ```
+
```
Comment (geoffrey45) [inline content/articles/lru-cache-implementation-in-javascript/index.md:90]: Remove this.

File: content/articles/lru-cache-implementation-in-javascript/index.md [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+The best data structure to use when implementing LRU cache is Double Linked List and Hash Table. The time complexity for this implementation is O(1). You can go ahead and try it out on your own as well.
+
+Happy Coding! :-)
+
+### References
+
+https://leetcode.com/problems/lru-cache/
+
+https://unsplash.com/photos/m_HRfLhgABo?utm_source=unsplash&utm_medium=referral&utm_content=creditShareLink
```
Comment (geoffrey45) [inline content/articles/lru-cache-implementation-in-javascript/index.md:135]: Remove this image.

File: content/articles/lru-cache-implementation-in-javascript/index.md [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+
+### Conclusion
+
+The best data structure to use when implementing LRU cache is Double Linked List and Hash Table. The time complexity for this implementation is O(1). You can go ahead and try it out on your own as well.
+
+Happy Coding! :-)
+
+### References
+
+https://leetcode.com/problems/lru-cache/
```
Comment (geoffrey45) [inline content/articles/lru-cache-implementation-in-javascript/index.md:133]: Use the correct syntax highlighting for markdown links.

```md
[LRU Cache - LeetCode](https://leetcode.com/problems/lru-cache/)
```

**Review by geoffrey45 — CHANGES_REQUESTED** [review]
@erixbiz Please check the comments below.

File: content/articles/lru-cache-implementation-in-javascript/index.md (hunk 4 above)
Comment (erixbiz) [reply content/articles/lru-cache-implementation-in-javascript/index.md:133]: I think this should be it having counter-checked with the documentation. An example I got from the documentation is :-
My favorite search engine is [Duck Duck Go](https://duckduckgo.com).

Comment (erixbiz) [reply content/articles/lru-cache-implementation-in-javascript/index.md:133]: [LRU Cache - LeetCode](https://leetcode.com/problems/lru-cache/)

Comment (erixbiz) [reply content/articles/lru-cache-implementation-in-javascript/index.md:133]: I think this should be it having counter checked with the documentation. An example I got from the documentation is 
My favorite search engine is [Duck Duck Go](https://duckduckgo.com).

Comment (geoffrey45) [reply content/articles/lru-cache-implementation-in-javascript/index.md:133]: @erixbiz The suggested syntax is meant to give your link a description.

File: content/articles/lru-cache-implementation-in-javascript/index.md [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+
+Still using the example red, green, white and blue, we needed to add black. Since it wasn't in our list, what we did was remove the oldest element blue and shift the entire list items.
+
+In a doubly-linked list, all you need to do is to remove the rear item and update the rear element to the previous element. Then add a new node and make it point to the second element and update the front. The entire operation is 0(1) time complexity.
+
+This will be clearly seen in the code example below.
+
+We can also optimize searching. How?
+
+Still using our previous example of cache size 4, with `[red, green, white, blue]`. searching for black. We can optimize our search by using a hash table. With a Hash, you will have a key and a value pair. The Key in this case will be the actual values in our cache. The value will be the address of our value as shown in the diagram below.
```
Comment (geoffrey45) [inline content/articles/lru-cache-implementation-in-javascript/index.md:51]: This whole section is still redundant. You just repeated what you said in the first section: What is LRU Cache?

 Please revisit this section removing all the redundancy.

File: content/articles/lru-cache-implementation-in-javascript/index.md [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+
+[Hash Table & Double Linked List Diagram](/ddl_hash.png)
+
+To add black to our list, we will remove the rear element blue from our list. Blue in this case is our least recently accessed item. Update the rear element to point to the previous element white. In our hash table, we will remove the value(address) that is associated with our key blue. Therefore, the value for our key blue will be null. A new node will is created where we put the black. The previous front gets updated to point to the node that has black. Black will also have an address associated with it as well.
+
+The shifting, in this case, is done in 0(1).
+
```
Comment (geoffrey45) [inline content/articles/lru-cache-implementation-in-javascript/index.md:58]: Please explain what is time complexity.

File: content/articles/lru-cache-implementation-in-javascript/index.md [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+3.  Put operation. The put operation will update the value of the key if it is found. If found, add the `Key` and the value pair to the cache. If the number of keys has exceeded the initialized capacity of the cache, evict the least recently accessed item.
+
+```javascript
+LRUCache.prototype.put = function (key, value) {
+
+    if (this.get(key) !== -1) { // if key does not exist, update last element value
+        this.tail.prev.value = value;
+    } else {
+        //need to check if map size is at capacity
+        if (this.map.size === this.capacity) {
```
Comment (geoffrey45) [inline content/articles/lru-cache-implementation-in-javascript/index.md:111]: Provide instructions on how to run this code. Also explain what to expect after running the code.

**Review by geoffrey45 — CHANGES_REQUESTED** [review]
@erixbiz Please check the comments below

File: content/articles/lru-cache-implementation-in-javascript/index.md (hunk 4 above)
Comment (erixbiz) [reply content/articles/lru-cache-implementation-in-javascript/index.md]: Capture
Sure. It shows that when I preview this file

**Comment by erixbiz** [comment]
@geoffrey45 done with the requested changes. Thank you

File: content/articles/lru-cache-implementation-in-javascript/index.md (hunk 4 above)
Comment (erixbiz) [reply content/articles/lru-cache-implementation-in-javascript/index.md]: I have identified the problem and corrected. Thanks

File: content/articles/lru-cache-implementation-in-javascript/index.md [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+
+Before adding `black`, we need to search first and see if it is available in the cache. To search for black, we go to the hash table. In the Hash table, key `black` will not have a corresponding value attached to it. This means that it is not present in our cache. This allows us to find out whether an item is in the cache with just O(1) time complexity.
+
+### Implementing LRU Cache in JavaScript.
+
+- Visit [leetecode](https://leetcode.com/)
+- Sign in to the account
+- Visit [lru-cache](https://leetcode.com/problems/lru-cache/)
+- Go through the problem statement.
+- Implement the lrucache class.
```
Comment (geoffrey45) [inline content/articles/lru-cache-implementation-in-javascript/index.md:70]: Provide instructions on how to run the code in this article without Signing up at LeetCode. Use the browser console to guide the reader on running the code. Only leave a link to the LeetCode question.

File: content/articles/lru-cache-implementation-in-javascript/index.md [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+};
+```
+Once done, click on the run code button. It is located on the bottom right. Below is the explanation of the expected output.
+
+Example
+
+#### Input
+```javascript
+["LRUCache", "put", "put", "get", "put", "get", "put", "get", "get", "get"]
+
```
Comment (geoffrey45) [inline content/articles/lru-cache-implementation-in-javascript/index.md:151]: Everything in this section and the ones below it are copied verbatim from the LeetCode challenge. Please remove them and use your own words to explain and guide the reader on how to run the code.

**Use step by step instructions, explaining what is happening and what is expected in each stage**

File: content/articles/lru-cache-implementation-in-javascript/index.md [hunk 10]
```diff
@@ -0,0 +1,192 @@
...
+lRUCache.get(1);    // return -1 (not found)
+
+lRUCache.get(3);    // return 3
+
+lRUCache.get(4);    // return 4
+```
+
+![Output](https://user-images.githubusercontent.com/84809276/133435896-f3a34988-fabd-43e2-8071-4fec1f93fede.PNG)
+
```
Comment (geoffrey45) [inline content/articles/lru-cache-implementation-in-javascript/index.md:185]: Download this image and place it in the article folder. Update this reference.

**Review by geoffrey45 — CHANGES_REQUESTED** [review]

**Comment by erixbiz** [comment]
@geoffrey45 done

**Comment by ahmadmardeni1** [comment]
plagiarism check done ✅

But there are two articles. Please delete the second one. @erixbiz

**Comment by geoffrey45** [comment]
@erixbiz Please remove the load balancing article from this pull request.

**Comment by erixbiz** [comment]
@geoffrey45  done

**Comment by geoffrey45** [comment]
> @geoffrey45 done

There's no changes made to this pull request. Please check the files changed tab.

**Comment by erixbiz** [comment]
@geoffrey45  thanks. Should be okay now.

**Comment by geoffrey45** [comment]
> @geoffrey45  thanks. Should be okay now. 

@erixbiz There's still two more files not removed. Please recheck your files.

**Comment by erixbiz** [comment]
@geoffrey45  done. thanks

**Review by geoffrey45 — APPROVED** [review]

**Comment by WanjaMIKE** [comment]
Your first two paragraphs appear to be the same. 

```
RAM has less space compared to the disk. This is why caching algorithms such as LRU can help invalidate entries that have not been used recently in the RAM.

The cache holds data in RAM. This makes the retrieval of data much faster than typical databases where data is stored on a disk. RAM has less space compared to the disk. This is why caching algorithms, such as least recently used (LRU) can help invalidate entries that have not been used recently in the RAM.
```
Such issues affect the quality of the article. Please address it and any other issues. Thanks

**Comment by erixbiz** [comment]
@WanjaMIKE done

**Pull request merged**